### PR TITLE
Develop. Transfer to Spring Boot.

### PR DESCRIPTION
**Pull request by RYelzhan**
Transformed application into sprig boot one. Added actuators with custom health indicators, as well as custom prometheus metrics. Added support for different environments (imaginary). Covered code with unit tests.

### CODE REVIEW

**Review by Atabek04 — APPROVED** [review]
Great job Yelzhan ! 
Please check the comments which I left for some minor improvements. Everything looks good overall—approved for submission.

File: src/main/java/com/epam/wca/gym/SpringGymApplication.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+@OpenAPIDefinition(
+        info = @Info(
+                title = "My Gym API",
+                description = "API for GYM management",
+                version = "2.1"
+        ),
+        servers = {
+                @Server(
+                        url = "http://localhost:8080",
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/SpringGymApplication.java:19]: Thanks, done.

File: src/main/java/com/epam/wca/gym/advice/GlobalExceptionHandler.java [hunk 2]
```diff
@@ -35,12 +45,30 @@ public ResponseEntity<String> handleForbiddenActionException(ForbiddenActionExce
...
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<String> handleInvalidLoginAttempt(IllegalArgumentException ex) {
+        return new ResponseEntity<>(ex.getMessage(), HttpStatus.BAD_REQUEST);
+    }
+/*
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<String> handleServerError() {
+        return new ResponseEntity<>(SERVER_ERROR_MESSAGE, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
+ */
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/advice/GlobalExceptionHandler.java:73]: Yeah, I added TODO: . Thanks.

File: src/main/java/com/epam/wca/gym/metrics/RequestCounterMetrics.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package com.epam.wca.gym.metrics;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import org.springframework.stereotype.Component;
+
+@Component
+public class RequestCounterMetrics {
+    private static final String requestCounterName = "request_counter";
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/metrics/RequestCounterMetrics.java:9]: Yeah, forgot about that. Thanks, done.

File: src/main/java/com/epam/wca/gym/service/impl/ProfileServiceImpl.java [hunk 4]
```diff
@@ -3,27 +3,23 @@
...
 import java.security.SecureRandom;
 
 @Slf4j
 @Service
+@RequiredArgsConstructor
 public class ProfileServiceImpl implements ProfileService {
-    private final UsernameDAO usernameDAO;
+    private final UsernameRepository usernameRepository;
     private static final String CHARACTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
     private static final int PASSWORD_LENGTH = 10;
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/ProfileServiceImpl.java:21]: Yeah, it does become easier to configure this way. Moved to application.properties. Thanks. Done.

File: src/main/java/com/epam/wca/gym/service/impl/AuthServiceImpl.java [hunk 5]
```diff
@@ -39,4 +43,13 @@ public User authenticate(String authHeader) {
         }
         throw new IllegalArgumentException("Not authenticated");
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/AuthServiceImpl.java:44]: Yeah, it is more appropriate approach. Done

File: src/main/java/com/epam/wca/gym/advice/GlobalExceptionHandler.java [hunk 6]
```diff
@@ -13,18 +16,25 @@
...
 import java.util.Map;
 
+@Slf4j
 @RestControllerAdvice
 public class GlobalExceptionHandler {
-    @ExceptionHandler(MyValidationException.class)
-    public ResponseEntity<String> handleMethodArgumentValidationExceptions(MyValidationException ex) {
+
+    private static final String INVALID_DATE_FORMAT = "Invalid date format. Please use the correct format (e.g., %s ).";
+    private static final String SERVER_ERROR_MESSAGE = "Error happened on server side.";
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/advice/GlobalExceptionHandler.java:24]: Hahahah, yeah. It seems like it. Thanks. Done.

File: src/main/java/com/epam/wca/gym/advice/GlobalExceptionHandler.java [hunk 7]
```diff
@@ -35,12 +45,30 @@ public ResponseEntity<String> handleForbiddenActionException(ForbiddenActionExce
 
     @ExceptionHandler(HttpMessageNotReadableException.class)
     public ResponseEntity<String> handleHttpMessageNotReadable() {
-
-        String message = "Invalid date format." +
-                " Please use the correct format (e.g., " +
-                AppConstants.DEFAULT_DATE_FORMAT +
-                " ).";
+        String message = INVALID_DATE_FORMAT.formatted(AppConstants.DEFAULT_DATE_FORMAT);
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/advice/GlobalExceptionHandler.java:48]: Yeah, it looks better. Thanks. Done

File: src/main/java/com/epam/wca/gym/config/SwaggerConfig.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    private String serverUrl;
+    @Value("${gym.api.server.description}")
+    private String serverDescription;
+
+    @Bean
+    public OpenAPI customOpenAPI() {
+        io.swagger.v3.oas.models.servers.Server server =
+                new io.swagger.v3.oas.models.servers.Server()
+                        .url(serverUrl)
+                        .description(serverDescription);
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/config/SwaggerConfig.java:39]: Can we use just `Server` and import package instead?

File: src/main/java/com/epam/wca/gym/controller/TraineeController.java [hunk 9]
```diff
@@ -3,169 +3,213 @@
...
 public class TraineeController {
-    @NonNull
-    private TraineeService traineeService;
-    @NonNull
-    private TrainerService trainerService;
-
-    @GetMapping("/profile/{username}")
+    private final TraineeService traineeService;
+    @Value("${gym.api.request.attribute.user}")
+    private String authenticatedUserRequestAttributeName;
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/controller/TraineeController.java:44]:
```suggestion
    private final TraineeService traineeService;
    
    @Value("${gym.api.request.attribute.user}")
    private String authenticatedUserRequestAttributeName;
```

File: src/main/java/com/epam/wca/gym/exception/AuthenticationException.java [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package com.epam.wca.gym.exception;
+
+public class AuthenticationException extends RuntimeException{
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/exception/AuthenticationException.java:3]: Please apply code formatting.

File: src/main/java/com/epam/wca/gym/exception/BadControllerRequestException.java [hunk 11]
```diff
@@ -0,0 +1,7 @@
+package com.epam.wca.gym.exception;
+
+public class BadControllerRequestException extends RuntimeException{
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/exception/BadControllerRequestException.java:3]: Please apply code formatting.

File: src/main/java/com/epam/wca/gym/exception/ControllerValidationException.java [hunk 12]
```diff
@@ -0,0 +1,7 @@
+package com.epam.wca.gym.exception;
+
+public class ControllerValidationException extends RuntimeException{
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/exception/ControllerValidationException.java:3]: Please apply code formatting.

File: src/main/java/com/epam/wca/gym/exception/InternalErrorException.java [hunk 13]
```diff
@@ -0,0 +1,7 @@
+package com.epam.wca.gym.exception;
+
+public class InternalErrorException extends RuntimeException{
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/exception/InternalErrorException.java:3]: Please apply code formatting.

File: src/main/java/com/epam/wca/gym/exception/ProfileNotFoundException.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package com.epam.wca.gym.exception;
+
+public class ProfileNotFoundException extends RuntimeException{
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/exception/ProfileNotFoundException.java:3]: Please apply code formatting.

File: src/main/java/com/epam/wca/gym/repository/TraineeRepository.java [hunk 15]
```diff
@@ -0,0 +1,10 @@
+package com.epam.wca.gym.repository;
+
+import com.epam.wca.gym.entity.Trainee;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/repository/TraineeRepository.java:7]: No need to use the `@Repository` annotation if you work with the `JpaRepository` interface. Please apply this to the entire project.

File: src/main/java/com/epam/wca/gym/service/impl/AuthServiceImpl.java [hunk 16]
```diff
@@ -12,31 +15,47 @@
...
 
-            if (values.length == 2) {
-                String username = values[0];
-                String password = values[1];
+            if (values.length == CREDENTIALS_LENGTH) {
+                String username = values[USERNAME_INDEX];
+                String password = values[PASSWORD_INDEX];
 
-                User user = userService.findByUniqueNameForAuthentication(username);
+                User user = userRepository.findUserByUserName(username);
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/service/impl/AuthServiceImpl.java:43]: Should we extract some utility logic to a standalone class or method to strengthen the Single Responsibility Principle?

File: src/main/java/com/epam/wca/gym/service/impl/TraineeServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    @Override
+    @Transactional
+    public TraineeSendDTO update(Trainee trainee, TraineeUpdateDTO traineeUpdateDTO) {
+        trainee.setFirstName(traineeUpdateDTO.firstName());
+        trainee.setLastName(traineeUpdateDTO.lastName());
+        trainee.setDateOfBirth(traineeUpdateDTO.dateOfBirth());
+        trainee.setAddress(traineeUpdateDTO.address());
+        trainee.setActive(traineeUpdateDTO.isActive());
+
+        // оказываеться finByUserName detaches object
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/service/impl/TraineeServiceImpl.java:70]: Please use English comments for the entire project.

File: src/test/java/com/epam/wca/gym/service/deprecated/AuthServiceImplTest.java [hunk 18]
```diff
@@ -21,7 +20,7 @@
 @ExtendWith(MockitoExtension.class)
 public class AuthServiceImplTest {
```
Comment (maximelyashevich) [inline src/test/java/com/epam/wca/gym/service/deprecated/AuthServiceImplTest.java:21]: Please use package-private access for your test classes and test methods if you work with JUnit 5.

File: src/main/java/com/epam/wca/gym/controller/UserController.java [hunk 19]
```diff
@@ -1,120 +1,133 @@
...
+//
+//    private final TrainingTypeDAO trainingTypeDAO;
+//    @Override
+//    public boolean isValid(String trainingType, ConstraintValidatorContext context) {
+//        if (trainingType == null || trainingType.trim().isEmpty()) {
+//            return true;
+//        }
+//        return trainingTypeDAO.findByName(trainingType.toUpperCase()).isPresent();
+//    }
+//}
```
Comment (maximelyashevich) [inline src/main/java/com/epam/wca/gym/controller/UserController.java:133]:
```suggestion
```

**Review by maximelyashevich — APPROVED** [review]
Reviewed from my side, please address my comment within whole project. Excellent work!

File: src/main/java/com/epam/wca/gym/config/SwaggerConfig.java (hunk 8 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/config/SwaggerConfig.java:39]: Yeah, done. Idea must have decided that this otion is better.

File: src/main/java/com/epam/wca/gym/exception/AuthenticationException.java (hunk 10 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/exception/AuthenticationException.java:3]: Done.

File: src/main/java/com/epam/wca/gym/exception/BadControllerRequestException.java (hunk 11 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/exception/BadControllerRequestException.java:3]: Done

File: src/main/java/com/epam/wca/gym/repository/TraineeRepository.java (hunk 15 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/repository/TraineeRepository.java:7]: Done.

File: src/main/java/com/epam/wca/gym/service/impl/AuthServiceImpl.java (hunk 16 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/AuthServiceImpl.java:43]: Yeah. Done. Code does look better and easier to comprohend this way.

File: src/main/java/com/epam/wca/gym/service/impl/TraineeServiceImpl.java (hunk 17 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/TraineeServiceImpl.java:70]: Ok, it was a momentary thing. Fixed.

File: src/main/java/com/epam/wca/gym/exception/ControllerValidationException.java (hunk 12 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/exception/ControllerValidationException.java:3]: Fixed.

File: src/main/java/com/epam/wca/gym/exception/ProfileNotFoundException.java (hunk 14 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/exception/ProfileNotFoundException.java:3]: Fixed.

File: src/main/java/com/epam/wca/gym/exception/InternalErrorException.java (hunk 13 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/exception/InternalErrorException.java:3]: Fixed.

File: src/main/java/com/epam/wca/gym/controller/TraineeController.java (hunk 9 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/TraineeController.java:44]: Fixed.

File: src/test/java/com/epam/wca/gym/service/deprecated/AuthServiceImplTest.java (hunk 18 above)
Comment (RYelzhan) [reply src/test/java/com/epam/wca/gym/service/deprecated/AuthServiceImplTest.java:21]: Fixed. Changed to default one.

File: src/main/java/com/epam/wca/gym/service/impl/AuthServiceImpl.java [hunk 20]
```diff
@@ -30,7 +34,7 @@ public User authenticate(String authHeader) {
                 String username = values[0];
                 String password = values[1];
 
-                User user = userService.findByUniqueNameForAuthentication(username);
+                User user = userRepository.findUserByUserName(username);
 
                 if (user != null && user.getPassword().equals(password)) {
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/AuthServiceImpl.java]: Added to TODO list this part. Thanks!

File: src/main/java/com/epam/wca/gym/aop/ApiTimingAspect.java [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+                .tag("method", methodName)
+                .description("Average Api Response Time")
+                .register(meterRegistry);
+
+        long start = System.nanoTime();
+
+        try {
+            return pjp.proceed();
+        } catch (Throwable e) {
+            throw new RuntimeException(e);
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/aop/ApiTimingAspect.java:43]: Added to TODO list to figure out what exception to throw.

File: src/main/java/com/epam/wca/gym/service/impl/TraineeServiceImpl.java [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+                                ". No Trainers Added");
+                    }
+
+                    addedTrainers.add(trainer);
+                });
+
+        trainee.addTrainers(addedTrainers);
+
+        // оказываеться finByUserName detaches object
+        traineeRepository.save(trainee);
```
[the remark this answers is not in the document]
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/service/impl/TraineeServiceImpl.java]: Thanks for advice. Done.

File: src/main/java/com/epam/wca/gym/controller/UserController.java (hunk 19 above)
Comment (RYelzhan) [reply src/main/java/com/epam/wca/gym/controller/UserController.java:133]: Done.

**Pull request merged**